text/3750-cfg-os-version-min.md

```diff
@@ -0,0 +1,130 @@
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Each supported platform will need to implement version string parsing logic (or re-use some provided defaults), maintain the logic in response to future changes, and update any version alias tables.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The overall mechanism proposed here builds on other well established primitives in Rust such as `cfg`.
+A mechanism which tries to bridge cross-platform differences under one `min_target_api_version` predicate [was suggested](https://github.com/rust-lang/rfcs/blob/b0f94000a3ddbd159013e100e48cd887ba2a0b54/text/0000-min-target-api-version.md) but was rejected due to different platforms having divergent needs.
+
+For many platforms, the `target_os` name and the `os_version_min` name will be identical.
+Even platforms that have multiple possible `versions` relevant to the OS will still have one primary version.
+E.g. for `linux` the primary version would refer to the kernel with `libc` being a secondary OS library version.
+Therefore it would be possible to simplify the syntax for the primary target OS version.
+E.g.: `cfg(target_os("macos", min_version = "..."))` or by having `os_version_min("macos", "...")` imply `#[cfg(target_os = "macos")]`.
+This means we'd need a more general syntax for `libc` and potentially other versioned libraries where the target OS is ambiguous.
+
+# Prior art
+[prior-art]: #prior-art
+
+In C it's common to be able to target different versions based on a preprocessor macro.
+For example, on Windows `WINVER` can be used:
+
+```c
+// If the minimum version is at least Windows 10
+#if (WINVER >= _WIN32_WINNT_WIN10)
+// ...
+#endif
+```
+
+This RFC is largely a version of [RFC #3379](https://github.com/rust-lang/rfcs/pull/3379) more narrowly scoped to just the most minimal lang changes.
+That RFC was in turn an updated version of [this RFC draft](https://github.com/rust-lang/rfcs/pull/3036), with the changes reflecting conversations from the draft review process and [further Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/CFG.20OS.20Redux.20.28migrated.29/near/294738760).
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Custom targets usually specify their configurations in JSON files.
+It is unclear how the target maintainers would add functions, types, and version compatibility information to these files.
+
+What exactly should the syntax be?
+Should we draw a distinction between cases where the `os_version_min` directly implies a specific `target_os` and cases where it doesn't (see alternatives)?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+The compiler could allow setting a higher minimum OS version than the target's default.
```

Comment:
Remark: I'll add that [`rustc` already has a mechanism for doing exactly this on Apple platforms](https://doc.rust-lang.org/1.83.0/rustc/platform-support/apple-darwin.html#os-version), namely the `*_DEPLOYMENT_TARGET` family of environment variables.